tgui/packages/tgui/components/MenuBarDropdown.tsx

```diff
@@ -0,0 +1,208 @@
+/**
+ * @file
+ * @copyright 2020 Aleksej Komarov
+ * @license MIT
+ */
+
+import { classes } from 'common/react';
+import { Component, createRef, InfernoNode, InfernoNodeArray, RefObject } from 'inferno';
+import { Box } from './Box';
+import { logger } from '../logging';
+import { Icon } from './Icon';
+
+type MenuProps = {
+  open: boolean;
+  options: InfernoNodeArray;
+  width: string;
+  menuRef: RefObject<HTMLElement>;
+  onOutsideClick: () => void;
+};
+
+class Menu extends Component<MenuProps> {
+  private readonly handleClick: (event) => void;
+
+  constructor(props) {
+    super(props);
+    this.handleClick = (event) => {
+      if (!this.props.menuRef.current) {
+        logger.log(`Menu.handleClick(): No ref`);
+        return;
+      }
+
+      if (this.props.menuRef.current.contains(event.target)) {
+        logger.log(`Menu.handleClick(): Inside`);
+      } else {
+        logger.log(`Menu.handleClick(): Outside`);
+        this.props.onOutsideClick();
+      }
+    };
+  }
+
+  // eslint-disable-next-line react/no-deprecated
+  componentWillMount() {
+    window.addEventListener('click', this.handleClick);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('click', this.handleClick);
+  }
+
+  render() {
+    const { open, options, width } = this.props;
+    const menu = open ? (
+      <div
+        className={'MenuBar__menu'}
+        style={{
+          width: width,
+        }}>
+        {options.length ? options : 'No Options Found'}
+      </div>
+    ) : null;
+
+    return { menu };
+  }
+}
+
+type MenuBarDropdownProps = {
+  open: boolean;
+  openWidth: string;
+  options: InfernoNodeArray;
+  disabled?: boolean;
+  display: InfernoNode;
+  onMouseOver: () => void;
+  onClick: () => void;
+  onOutsideClick: () => void;
+  className?: string;
+};
+
+class MenuBarButton extends Component<MenuBarDropdownProps> {
+  private readonly menuRef: RefObject<HTMLDivElement>;
+
+  constructor(props) {
+    super(props);
+    this.menuRef = createRef();
+  }
+
+  render() {
+    const { props } = this;
+    const {
+      open,
+      openWidth,
+      options,
+      disabled,
+      display,
+      onMouseOver,
+      onClick,
+      onOutsideClick,
+      ...boxProps
+    } = props;
+    const { className, ...rest } = boxProps;
+
+    return (
+      <div ref={this.menuRef} className="MenuBar">
+        <Box
+          className={classes(['MenuBar__MenuBarButton', className])}
+          {...rest}
+          onClick={disabled ? undefined : onClick}
+          onmouseover={onMouseOver}>
+          <span className="MenuBar__selected-text">{display}</span>
+        </Box>
+        {open && (
+          <Menu
+            open={open}
```

Comment:
The open property seems pointless when you already hide the entire node using a boolean expression just above.